File: contest/LeetCode_Weekly_Contest_204/Maximum_Length_of_Subarray_With_Positive_Product.java
import static java.lang.Integer.max;
class Solution {
    public int getMaxLen(int[] nums) {
        int n = nums.length;
        int L = 0, R = 0, ret = 0;
        while (L < n) {
            while (R < n && nums[R] == 0) R++;
            L = R;
            int neg = 0;
            while (R < n && nums[R] != 0) {
                if (nums[R] < 0) neg++;
                if ((neg & 1) == 0) ret = max(ret, R - L + 1);
                R++;
            }
            while (L < R) {
                if ((neg & 1) == 0) ret = max(ret, R - L);
                if (nums[L] < 0) neg--;
                L++;
            }
        }
        return ret;
    }
}

File: contest/LeetCode_Weekly_Contest_204/Maximum_Length_of_Subarray_With_Positive_Product.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var getMaxLen = function(nums) {
    var n = nums.length;
    var L = 0, R = 0, ret = 0;
    while (L < n) {
        while (R < n && nums[R] === 0) R++;
        L = R;
        var neg = 0;
        while (R < n && nums[R] !== 0) {
            if (nums[R] < 0) neg++;
            if (!(neg & 1)) ret = Math.max(ret, R - L + 1); 
            R++;
        }
        while (L < R) {
            if (!(neg & 1)) ret = Math.max(ret, R - L);
            if (nums[L] < 0) neg--;
            L++;
        }
    }
    return ret;
};